Excluding project from affected:build makes the build fail for that project

## Expected Behavior

Command:
```
nx affected:build --all --prod --exclude=foo --vendor-chunk --no-progress
```
Running this command builds all apps except for `foo` and exits with success.

## Current Behavior

The command builds every app except for `foo`, but then exists with the following error about `foo`:

```
Running build for affected projects failed.
Failed projects: foo
```

### Context

- version of Nx used: 7.8.0
- version of Angular CLI used: ~7.3.5
- version of Angular DevKit used: ~0.13.0

### Failure Logs

Here's the full redacted log:

```
Running build for affected projects failed.
Failed projects: rest-simulator
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! myMonorepo@0.0.0-semantically-released affected:build: `nx affected:build "--all" "--prod" "--exclude=foo" "--vendor-chunk" "--no-progress"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the myMonorepo@0.0.0-semantically-released affected:build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/williamghelfi/.npm/_logs/2019-04-17T09_01_40_636Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! myMonorepo@0.0.0-semantically-released build:ci: `npm run affected:build -- --all --prod --exclude=foo --vendor-chunk --no-progress`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the myMonorepo@0.0.0-semantically-released build:ci script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/williamghelfi/.npm/_logs/2019-04-17T09_01_40_668Z-debug.log
```

--- a/packages/workspace/src/command-line/affected.ts
+++ b/packages/workspace/src/command-line/affected.ts
@@ -20,7 +20,6 @@ import { DefaultReporter } from '../tasks-runner/default-reporter';
 export function affected(command: string, parsedArgs: yargs.Arguments): void {
   const { nxArgs, overrides } = splitArgsIntoNxArgsAndOverrides(parsedArgs);
 
-  const env = readEnvironment(nxArgs.target);
   const projectGraph = createProjectGraph();
   let affectedGraph = nxArgs.all
     ? projectGraph
@@ -33,11 +32,13 @@ export function affected(command: string, parsedArgs: yargs.Arguments): void {
       withDeps(projectGraph, Object.values(affectedGraph.nodes))
     );
   }
-  const affectedProjects = Object.values(
-    parsedArgs.all ? projectGraph.nodes : affectedGraph.nodes
-  )
+  const projects = parsedArgs.all ? projectGraph.nodes : affectedGraph.nodes;
+  const env = readEnvironment(nxArgs.target, projects);
+  const affectedProjects = Object.values(projects)
     .filter(n => !parsedArgs.exclude.includes(n.name))
-    .filter(n => !parsedArgs.onlyFailed || !env.workspace.getResult(n.name));
+    .filter(
+      n => !parsedArgs.onlyFailed || !env.workspaceResults.getResult(n.name)
+    );
 
   try {
     switch (command) {

--- a/packages/workspace/src/command-line/run-many.ts
+++ b/packages/workspace/src/command-line/run-many.ts
@@ -14,9 +14,13 @@ import { DefaultReporter } from '../tasks-runner/default-reporter';
 
 export function runMany(parsedArgs: yargs.Arguments): void {
   const { nxArgs, overrides } = splitArgsIntoNxArgsAndOverrides(parsedArgs);
-  const env = readEnvironment(nxArgs.target);
   const projectGraph = createProjectGraph();
   const projects = projectsToRun(nxArgs, projectGraph);
+  const projectMap: Record<string, ProjectGraphNode> = {};
+  projects.forEach(proj => {
+    projectMap[proj.name] = proj;
+  });
+  const env = readEnvironment(nxArgs.target, projectMap);
   runCommand(
     projects,
     projectGraph,

--- a/packages/workspace/src/command-line/run-one.ts
+++ b/packages/workspace/src/command-line/run-one.ts
@@ -9,9 +9,11 @@ export function runOne(opts: {
   configuration: string;
   overrides: any;
 }): void {
-  const env = readEnvironment(opts.target);
   const projectGraph = createProjectGraph();
   const projects = [projectGraph.nodes[opts.project]];
+  const env = readEnvironment(opts.target, {
+    [opts.project]: projectGraph.nodes[opts.project]
+  });
   runCommand(
     projects,
     projectGraph,

--- a/packages/workspace/src/command-line/workspace-results.spec.ts
+++ b/packages/workspace/src/command-line/workspace-results.spec.ts
@@ -2,14 +2,21 @@ import * as fs from 'fs';
 
 import { WorkspaceResults } from './workspace-results';
 import { serializeJson } from '../utils/fileutils';
-import { stripIndents } from '@angular-devkit/core/src/utils/literals';
-import { output } from '../utils/output';
+import { ProjectType } from '..//core/project-graph';
 
 describe('WorkspacesResults', () => {
   let results: WorkspaceResults;
 
   beforeEach(() => {
-    results = new WorkspaceResults('test');
+    results = new WorkspaceResults('test', {
+      proj: {
+        name: 'proj',
+        type: ProjectType.app,
+        data: {
+          files: []
+        }
+      }
+    });
   });
 
   it('should be instantiable', () => {
@@ -80,7 +87,15 @@ describe('WorkspacesResults', () => {
         })
       );
 
-      results = new WorkspaceResults('test');
+      results = new WorkspaceResults('test', {
+        proj: {
+          name: 'proj',
+          type: ProjectType.app,
+          data: {
+            files: []
+          }
+        }
+      });
 
       expect(fs.readFileSync).toHaveBeenCalledWith('dist/.nx-results', 'utf-8');
       expect(results.getResult('proj')).toBe(false);
@@ -90,7 +105,15 @@ describe('WorkspacesResults', () => {
       spyOn(fs, 'readFileSync').and.returnValue('invalid json');
 
       const runTests = () => {
-        results = new WorkspaceResults('test');
+        results = new WorkspaceResults('test', {
+          proj: {
+            name: 'proj',
+            type: ProjectType.app,
+            data: {
+              files: []
+            }
+          }
+        });
       };
 
       expect(runTests).not.toThrow();
@@ -107,9 +130,41 @@ describe('WorkspacesResults', () => {
         })
       );
 
-      results = new WorkspaceResults('build');
+      results = new WorkspaceResults('build', {
+        proj: {
+          name: 'proj',
+          type: ProjectType.app,
+          data: {
+            files: []
+          }
+        }
+      });
 
       expect(results.getResult('proj')).toBeUndefined();
+    });
+
+    it('should invalidate existing results when the project is not run', () => {
+      spyOn(fs, 'readFileSync').and.returnValue(
+        serializeJson({
+          command: 'test',
+          results: {
+            proj: true,
+            proj2: false
+          }
+        })
+      );
+
+      results = new WorkspaceResults('build', {
+        proj: {
+          name: 'proj',
+          type: ProjectType.app,
+          data: {
+            files: []
+          }
+        }
+      });
+
+      expect(results.hasFailure).toEqual(false);
     });
   });
 });

--- a/packages/workspace/src/command-line/workspace-results.ts
+++ b/packages/workspace/src/command-line/workspace-results.ts
@@ -1,6 +1,7 @@
 import * as fs from 'fs';
 import { readJsonFile, writeJsonFile } from '../utils/fileutils';
 import { unlinkSync } from 'fs';
+import { ProjectGraphNode } from '../core/project-graph';
 
 const RESULTS_FILE = 'dist/.nx-results';
 
@@ -26,7 +27,10 @@ export class WorkspaceResults {
     return Object.values(this.commandResults.results).some(result => !result);
   }
 
-  constructor(private command: string) {
+  constructor(
+    private command: string,
+    private projects: Record<string, ProjectGraphNode>
+  ) {
     const resultsExists = fs.existsSync(RESULTS_FILE);
     this.startedWithFailedProjects = false;
     if (resultsExists) {
@@ -35,6 +39,7 @@ export class WorkspaceResults {
         this.startedWithFailedProjects = commandResults.command === command;
         if (this.startedWithFailedProjects) {
           this.commandResults = commandResults;
+          this.invalidateOldResults();
         }
       } catch {
         /**
@@ -60,5 +65,13 @@ export class WorkspaceResults {
 
   setResult(projectName: string, result: boolean) {
     this.commandResults.results[projectName] = result;
+  }
+
+  private invalidateOldResults() {
+    Object.keys(this.commandResults.results).forEach(projectName => {
+      if (!this.projects[projectName]) {
+        delete this.commandResults.results[projectName];
+      }
+    });
   }
 }

--- a/packages/workspace/src/core/file-utils.ts
+++ b/packages/workspace/src/core/file-utils.ts
@@ -199,12 +199,15 @@ export function readWorkspaceFiles(): FileData[] {
   return files;
 }
 
-export function readEnvironment(target: string): Environment {
+export function readEnvironment(
+  target: string,
+  projects: Record<string, ProjectGraphNode>
+): Environment {
   const nxJson = readNxJson();
   const workspaceJson = readWorkspaceJson();
-  const workspace = new WorkspaceResults(target);
+  const workspaceResults = new WorkspaceResults(target, projects);
 
-  return { nxJson, workspaceJson, workspace };
+  return { nxJson, workspaceJson, workspaceResults };
 }
 
 /**

--- a/packages/workspace/src/core/shared-interfaces.ts
+++ b/packages/workspace/src/core/shared-interfaces.ts
@@ -1,3 +1,5 @@
+import { WorkspaceResults } from '@nrwl/workspace/src/command-line/workspace-results';
+
 export type ImplicitDependencyEntry<T = '*' | string[]> = {
   [key: string]: T | ImplicitJsonSubsetDependency<T>;
 };
@@ -28,5 +30,5 @@ export interface NxJsonProjectConfig {
 export interface Environment {
   nxJson: NxJson;
   workspaceJson: any;
-  workspace: any;
+  workspaceResults: WorkspaceResults;
 }

--- a/packages/workspace/src/tasks-runner/run-command.ts
+++ b/packages/workspace/src/tasks-runner/run-command.ts
@@ -14,7 +14,7 @@ type RunArgs = yargs.Arguments & ReporterArgs;
 export async function runCommand<T extends RunArgs>(
   projectsToRun: ProjectGraphNode[],
   projectGraph: ProjectGraph,
-  { nxJson, workspace }: Environment,
+  { nxJson, workspaceResults }: Environment,
   nxArgs: NxArgs,
   overrides: any,
   reporter: any
@@ -53,11 +53,11 @@ export async function runCommand<T extends RunArgs>(
     next: (event: any) => {
       switch (event.type) {
         case AffectedEventType.TaskComplete: {
-          workspace.setResult(event.task.target.project, event.success);
+          workspaceResults.setResult(event.task.target.project, event.success);
           break;
         }
         case AffectedEventType.TaskCacheRead: {
-          workspace.setResult(event.task.target.project, event.success);
+          workspaceResults.setResult(event.task.target.project, event.success);
           cached.push(event.task.target.project);
           break;
         }
@@ -68,15 +68,15 @@ export async function runCommand<T extends RunArgs>(
       // fix for https://github.com/nrwl/nx/issues/1666
       if (process.stdin['unref']) (process.stdin as any).unref();
 
-      workspace.saveResults();
+      workspaceResults.saveResults();
       reporter.printResults(
         nxArgs,
-        workspace.failedProjects,
-        workspace.startedWithFailedProjects,
+        workspaceResults.failedProjects,
+        workspaceResults.startedWithFailedProjects,
         cached
       );
 
-      if (workspace.hasFailure) {
+      if (workspaceResults.hasFailure) {
         process.exit(1);
       }
     }

